### ISSUE TITLE
0.11.4 false positive

### ISSUE THREAD

**Issue by matthargett** [danielstjules/jsinspect#56]
0.11.4 is reporting the following duplication even though our threshold is set to 40 in the jsinspectrc. It wasn't reporting this false positive in 0.9:

```
Match - 2 instances

./App/Containers/GuestJoinLayout.js:168,168
</Animatable.View>

./App/Containers/SignInLayout.js:160,160
</Animatable.View>

------------------------------------------------------------
```

The relevant part of GuestJoinLayout.js:
```
 render () {
    return (
      <View style={Styles.container}>
        <View style={Styles.content}>
          <Animatable.View ref='content' animation='fadeInUp' duration={2000} ease='ease-in'>
            <LinearGradient
              start={{x: 0.0, y: 0.0}} end={{x: 0.0, y: 1.0}}
              locations={[0, 1]}
              colors={['#074595', '#6589A4']}
              style={Styles.signInContainer}
            >
              {this.renderTitle()}
              {this.renderMeetingId()}
              {this.renderPasscode()}
              {this.renderUsernameInput()}
              {this.renderDropdown()}
              {this.renderJoinButton()}
              {/* this.renderSignIn() */}
            </LinearGradient>
          </Animatable.View>
        </View>
      </View>
    )
  }
```

and SignInLayout.js:
```
render () {
    return (
      <Animatable.View ref='container' style={Styles.container} easing='ease-in'>
        <LinearGradient
          start={{x: 0.0, y: 0.0}} end={{x: 0.0, y: 1.0}}
          locations={[0, 1]}
          colors={['#074595', '#6589A4']}
          style={Styles.blurOverlay}>
          <View style={Styles.content}>
            {this.renderLogo()}
            {this.renderErrorMessage()}
            {this.renderUsernameInput()}
            {this.renderPasswordInput()}
            {this.renderNextButton()}
            {this.renderSignInOptions()}
            {this.renderSignUp()}
          </View>
        </LinearGradient>
      </Animatable.View>
    )
  }
```

**Comment by danielstjules** [comment]
I'm pretty sure jsinspect 0.9 and below just completely ignored any JSX and kept parsing the files. The tool was completely rewritten in [0.10](https://github.com/danielstjules/jsinspect/releases/tag/0.10.0) as I found people were mostly interested in copy-paste detection, and the API was pretty inconsistent. As a result, thresholds will likely need to be tweaked.

That said, that's definitely a bug! Looks like there's 2 issues that need to be addressed:
* End line for the sequence of JSX nodes is incorrect
* JSX ASTs tend to be quite verbose, will probably need to make them more compact so as to avoid tripping up at lower thresholds

Thanks for letting me know!

**Comment by danielstjules** [comment]
> End line for the sequence of JSX nodes is incorrect

Fixed in https://github.com/danielstjules/jsinspect/commit/20a16c9057fe2207d0ec1429ac92b8d62addb529 The issue is that `NodeUtils.getChildren` isn't guaranteeing children order. It worked for most nodes, e.g. function params get visited before its body, but not JSX. Will probably update the function to peak at node location before caching attribute order. [Previous versions of jsinspect hardcoded attribute traversal order](https://github.com/danielstjules/acorn-bfs/blob/b8072f70e552389982b8e8e67e22ce88750f14ac/index.js#L32-L257)... but it's not very future-proof.

> JSX ASTs tend to be quite verbose, will probably need to make them more compact so as to avoid tripping up at lower thresholds

This is still an issue, and this screenshot highlights what's being reported as common:

![screen shot 2017-03-27 at 6 50 12 pm](https://cloud.githubusercontent.com/assets/817212/24381574/fe683206-131e-11e7-8e1a-e335c7e653ca.png)

**Comment by danielstjules** [comment]
Quick explanation as to why it's not really a false positive, but a higher threshold might be necessary for JSX files. Let's look at this small snippet from the first example:

``` javascript
start={{x: 0.0, y: 0.0}} end={{x: 0.0, y: 1.0}}
    locations={[0, 1]}
```

This breaks down into the following 25 nodes
```
JSXIdentifier (start), JSXExpressionContainer, ObjectExpression, ObjectProperty, Identifier (x), NumericLiteral (0), ObjectProperty, Identifier (y), NumericLiteral (0), JSXAttribute, JSXIdentifier (end), JSXExpressionContainer, ObjectExpression, ObjectProperty, Identifier (x), NumericLiteral (0), ObjectProperty, Identifier (y), NumericLiteral (1), JSXAttribute, JSXIdentifier (locations), JSXExpressionContainer, ArrayExpression, NumericLiteral (0), NumericLiteral 1
```

**Comment by danielstjules** [comment]
And as a note, here's what would be reported after the latest patch release (note that it's not just a single line)

![screen shot 2017-03-28 at 1 49 48 pm](https://cloud.githubusercontent.com/assets/817212/24419553/81428c50-13bd-11e7-807a-46495f1f939f.png)

**Comment by danielstjules** [comment]
This has been fixed in https://github.com/danielstjules/jsinspect/commit/119e32daee8122e0288f6f0024bdbcd4939655f1

![screen shot 2017-03-28 at 2 48 36 pm](https://cloud.githubusercontent.com/assets/817212/24421924/bb803f68-13c5-11e7-9a3b-2eec3c51e97c.png)

This makes up 40 nodes:
```
'ease-in'>
  <LinearGradient
    start={{x: 0.0, y: 0.0}} end={{x: 0.0, y: 1.0}}
    locations={[0, 1]}
    colors={['#074595', '#6589A4']}
    style={Styles.
```

Setting the threshold to 42 will avoid catching this LinearGradient. Thanks!

**Comment by matthargett** [comment]
Awesome, thanks for the fast turnaround! :) :1st_place_medal:
